architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,442 @@
+
+### Architecture Diagram
+
+**For clarity, this design document will depict Dynamic RP as a separate component. We might choose to host it in UCP if that makes sense.**
+
+#### Resource Type Management
+
+![User-defined type registration](./2024-07-user-defined-types/udt-registration.png)
+
+<center>
+
+*Figure: Diagram of user-defined type registration* [link](https://excalidraw.com/#json=V8jLG-XK7VYZQ5AMAfsRJ,zwUXvuelafgFL2AAGZDsEA)
+
+</center>
+
+#### Request Path
+
+![User-defined type request path](./2024-07-user-defined-types/udt-request-path.png)
+
+<center>
+
+*Figure: Diagram of user-defined type request path* [link](https://excalidraw.com/#json=cqcrjP7qY1UgbWETA6nDC,bW1ZTLIrshMgCgD50NrzBg)
+
+</center>
+
+#### Asynchronous Processing
+
+![User-defined type asynchronous processing](./2024-07-user-defined-types/udt-async-processing.png)
+
+<center>
+
+*Figure: Diagram of user-defined type async processing* [link](https://excalidraw.com/#json=kHTPUX6ihLJOsW01eSL-o,tr8NQU3nHDBA2OfhGTWPzw)
+
+</center>
+
+
+### Detailed Design
+
+This is an overview design document and so the bulk of the decision making will be carried out via follow-up documents. The following documents are planned:
+
+- Resource Type registration APIs (TODO)
+	- Includes enhancements to UCP routing.
+- Declarative Resource Type schemas (TODO)
+	- Includes API design guidelines, naming guidelines.
+	- Includes validation behaviors.
+	- Includes capability model (eg: how does a resource type opt-in to Recipes or Connections)
+- Declarative API versioning (TODO)
+	- Includes support for multiple versions and support for breaking changes.
+- Dynamic RP (TODO)
+	- Ties together the decisions about validation, versioning, storage.
+	- Calls into Recipe engine.
+- Tooling support for using user-defined types (TODO)
+	- Includes new CLI commands for managing user-defined types.
+	- Includes updates to existing CLI commands for user-defined types.
+	- Includes updates to existing Dashboard functionality for user-defined types.
+- Tooling support for authoring user-defined types (TODO)
+	- Includes functionality for scaffolding a user-defined type project.
+	- Includes functionality for generating or authoring a resource type manifest.
+	- Includes functionality for generating a Bicep extension.
+- Community repository and process for resource-types. (TODO)
+
+----
+
+There are also a few major decisions to be made at a top-down level, which are captured here.
+
+#### Declarative User-defined types are most valuable
+
+Our scenario doc and jobs-to-be-done analysis highlights that authoring user-defined-types needs to be easy and have a low barrier to entry. This design prop supports that goal by providing a fully-declarative path for implementing a user-defined type.
+
+- Validation is declarative (OpenAPI + CEL with no imperative component).
```

Comment:
CEL is an expression/validation language designed to be embedded in server-side usecases like ours. It's been adopted by Kubernetes and is getting a lot of traction as a result. It's also not going to go away because it's got a major adopter.

https://kubernetes.io/docs/reference/using-api/cel/

Reply:
Based on this question I'll either remove this reference for clarity or spend more time explaining.

Reply:
I've cleaned this up to be less perscriptive. Also added an examples of when declarative functionality and imperative functionality are needed (webhooks). 

This doc won't be a comprehensive reference.